### PR TITLE
Feat #458 - Implement report designer ui

### PR DESCRIPTION
**Pull request by adrianchivu**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-IME-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-IME [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-IME-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### CODE REVIEW

File: CDP4IME/Shell.xaml [hunk 1]
```diff
@@ -54,11 +54,6 @@
 
             <dxr:RibbonControl.PageHeaderItems>
 
-                <dxb:BarButtonItem Name="bAbout"
```
Comment (alexatrhea) [inline CDP4IME/Shell.xaml:57]: Please revert this change. It does not do much for now, but hardly belongs in this PR.

Comment (adrianchivu) [reply CDP4IME/Shell.xaml:57]: Done

File: CDP4Composition/Mvvm/Behaviours/RibbonCategoryBehavior.cs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        /// </param>
+        private static void CategoryNameChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            if (d is RibbonCategoryBehavior ribbonCategoryBehavior)
+            {
+                ribbonCategoryBehavior.CategoryName = (string)e.NewValue;
+            }
+        }
+    }
+}
```
Comment (lxatrhea) [inline CDP4Composition/Mvvm/Behaviours/RibbonCategoryBehavior.cs:172]: Please end all code files with a line break.

File: CDP4Composition/Ribbon/ExtendedRibbonPageCategory.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        /// The page will only be added by the <see cref="RibbonAdapter"/> if the <see cref="RibbonPage"/> with the specified <see cref="RegionName"/>
+        /// property can be found in the <see cref="RibbonControl"/>
+        /// </remarks>
+        public string ContainerRegionName
+        {
+            get => (string)this.GetValue(ContainerRegionNameProperty);
+            set => this.SetValue(ContainerRegionNameProperty, value);
+        }
+    }
+}
```
Comment (lxatrhea) [inline CDP4Composition/Ribbon/ExtendedRibbonPageCategory.cs:64]: Please end all code files with a line break.

File: CDP4Reporting.Tests/Behaviour/ReportDesignerBehaviourTextFixture.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
+﻿// -------------------------------------------------------------------------------------------------
+// <copyright file="ReportDesignerBehaviourTextFixture.cs" company="RHEA System S.A.">
+//   Copyright (c) 2020 RHEA System S.A.
+// </copyright>
```
Comment (lxatrhea) [inline CDP4Reporting.Tests/Behaviour/ReportDesignerBehaviourTextFixture.cs:4]: Please add the appropriate header (AGPL)

File: CDP4Reporting.Tests/Behaviour/ReportDesignerBehaviourTextFixture.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+        [Test]
+        public void VerifyGetSetRibbonMergeCategoryName()
+        {
+            Assert.IsNull(reportDesignerBehaviour.RibbonMergeCategoryName);
+            reportDesignerBehaviour.RibbonMergeCategoryName = "abc";
+            Assert.AreEqual(reportDesignerBehaviour.RibbonMergeCategoryName, "abc");
+        }
+
+        [TearDown]
```
Comment (lxatrhea) [inline CDP4Reporting.Tests/Behaviour/ReportDesignerBehaviourTextFixture.cs:44]: TearDown method should be right under the SetUp method at the top of the file.

File: CDP4Reporting.Tests/Behaviour/RibbonCategoryBehaviorTextFixture.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
+﻿// -------------------------------------------------------------------------------------------------
+// <copyright file="RibbonCategoryBehaviorTextFixture.cs" company="RHEA System S.A.">
+//   Copyright (c) 2020 RHEA System S.A.
```
Comment (lxatrhea) [inline CDP4Reporting.Tests/Behaviour/RibbonCategoryBehaviorTextFixture.cs:3]: Please add the appropriate header (AGPL)

File: CDP4Reporting.Tests/CDP4ReportingModuleTestFixture.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
+﻿// -------------------------------------------------------------------------------------------------
+// <copyright file="CDP4ReportingModuleTestFixture.cs" company="RHEA System S.A.">
+//   Copyright (c) 2020 RHEA System S.A.
```
Comment (lxatrhea) [inline CDP4Reporting.Tests/CDP4ReportingModuleTestFixture.cs:3]: Please add the appropriate header (AGPL)

File: CDP4Reporting.Tests/Ribbon/ExtendedRibbonPageCategoryTextFixture.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
+﻿// -------------------------------------------------------------------------------------------------
+// <copyright file="ExtendedRibbonPageCategoryTextFixture.cs" company="RHEA System S.A.">
+//   Copyright (c) 2020 RHEA System S.A.
```
Comment (lxatrhea) [inline CDP4Reporting.Tests/Ribbon/ExtendedRibbonPageCategoryTextFixture.cs:3]: Please add the appropriate header (AGPL)

File: CDP4Reporting.Tests/Ribbon/ExtendedRibbonPageCategoryTextFixture.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        public void GetSetTheDefaultValueOfTheDependencyPropertyContainer()
+        {
+            Assert.IsNull(extendedRibbonPageCategory.ContainerRegionName);
+
+            extendedRibbonPageCategory.ContainerRegionName = "abc";
+
+            Assert.AreEqual(extendedRibbonPageCategory.ContainerRegionName, "abc");
+        }
+
+        [TearDown]
```
Comment (lxatrhea) [inline CDP4Reporting.Tests/Ribbon/ExtendedRibbonPageCategoryTextFixture.cs:46]: Please move the TearDown method right under the SetUp method.

File: CDP4Reporting.Tests/ViewModels/ReportingRibbonViewModelTextFixture.cs [hunk 10]
```diff
@@ -0,0 +1,149 @@
+﻿// -------------------------------------------------------------------------------------------------
+// <copyright file="ReportingRibbonViewModelTextFixture.cs" company="RHEA System S.A.">
+//   Copyright (c) 2020 RHEA System S.A.
```
Comment (lxatrhea) [inline CDP4Reporting.Tests/ViewModels/ReportingRibbonViewModelTextFixture.cs:3]: Please add the appropriate header (AGPL)

File: CDP4Reporting/Behaviours/ReportDesignerBehaviour.cs [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+        [ExcludeFromCodeCoverage]
+        protected override void OnDetaching()
+        {
+            this.AssociatedObject.Loaded -= this.Loaded;
+            this.AssociatedObject.Unloaded -= this.Unloaded;
+
+            base.OnDetaching();
+        }
+    }
+}
```
Comment (lxatrhea) [inline CDP4Reporting/Behaviours/ReportDesignerBehaviour.cs:203]: Please end all code files with a line break.

File: CDP4Reporting/CDP4Reporting.csproj [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+    <ProjectGuid>{C31FF150-E45C-4136-96CD-58FBF9B00C08}</ProjectGuid>
+    <TargetFramework>net452</TargetFramework>
+
+    <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\</SolutionDir>
+    <AssemblyTitle>CDP4Reporting</AssemblyTitle>
+    <Company>RHEA System S.A.</Company>
+    <Product>CDP4Reporting</Product>
+    <Description>CDP4 Reporting Plugin</Description>
+    <Copyright>Copyright © RHEA System S.A.</Copyright>
+    <AssemblyVersion>6.0.0.0</AssemblyVersion>
```
Comment (lxatrhea) [inline CDP4Reporting/CDP4Reporting.csproj:12]: Verions should be 0.0.1.0 because it is not finished yet. 
Reporting will probably ship as a alfa/beta version with the next IME release.

File: CDP4Reporting/ViewModels/ReportDesignerRibbonViewModel.cs [hunk 13]
```diff
@@ -0,0 +1,65 @@
+﻿// -------------------------------------------------------------------------------------------------
+// <copyright file="ReportingRibbonPageViewModel.cs" company="RHEA System S.A.">
+//   Copyright (c) 2020 RHEA System S.A.
```
Comment (lxatrhea) [inline CDP4Reporting/ViewModels/ReportDesignerRibbonViewModel.cs:3]: Please use appropriate file header (AGPL)

File: CDP4Reporting/ViewModels/ReportDesignerViewModel.cs [hunk 14]
```diff
@@ -0,0 +1,52 @@
+﻿// -------------------------------------------------------------------------------------------------
+// <copyright file="RelationshipEditorViewModel.cs" company="RHEA System S.A.">
+//   Copyright (c) 2015-2020 RHEA System S.A.
```
Comment (lxatrhea) [inline CDP4Reporting/ViewModels/ReportDesignerViewModel.cs:3]: Please use appropriate file header (AGPL)

File: CDP4Reporting/ViewModels/ReportDesignerViewModel.cs [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        /// <param name="thing">
+        /// The <see cref="Thing"/> to display
+        /// </param>
+        /// <param name="participant">The <see cref="Participant"/> that this open <see cref="Iteration"/> is tied to.</param>
+        /// <param name="session">
+        /// The session.
+        /// </param>
+        /// <param name="thingDialogNavigationService"></param>
+        /// <param name="panelNavigationService">The panel navigation service.</param>
+        public ReportDesignerViewModel(Iteration thing, Participant participant, ISession session, IThingDialogNavigationService thingDialogNavigationService, IPanelNavigationService panelNavigationService, IDialogNavigationService dialogNavigationService, IPluginSettingsService pluginSettingsService)
```
Comment (lxatrhea) [inline CDP4Reporting/ViewModels/ReportDesignerViewModel.cs:45]: I think the dialogNavigationService and pluginSettingsService should also be added to XmlDocs.

File: CDP4Reporting/Views/ReportDesigner.xaml.cs [hunk 16]
```diff
@@ -0,0 +1,68 @@
+﻿namespace CDP4Reporting.Views
```
Comment (lxatrhea) [inline CDP4Reporting/Views/ReportDesigner.xaml.cs:1]: Please add appropriate file header (AGPL)

File: CDP4Reporting/Views/ReportDesignerRibbon.xaml.cs [hunk 17]
```diff
@@ -0,0 +1,38 @@
+﻿// --------------------------------------------------------------------------------------------------------------------
+// <copyright file="ReportingRibbonPage.xaml.cs" company="RHEA System S.A.">
+//   Copyright (c) 2020 RHEA System S.A.
```
Comment (lxatrhea) [inline CDP4Reporting/Views/ReportDesignerRibbon.xaml.cs:3]: Please use appropriate file header (AGPL)

File: CDP4Reporting/Views/ReportDesignerRibbon.xaml.cs [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    /// Interaction logic for ObjectBrowserRibbon
+    /// </summary>
+    [Export(typeof(ReportDesignerRibbon))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public partial class ReportDesignerRibbon : ExtendedRibbonPageGroup, IView
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ReportDesignerRibbon"/> class.
+        /// </summary>
+        /// <param name="panelNavigationService">
```
Comment (lxatrhea) [inline CDP4Reporting/Views/ReportDesignerRibbon.xaml.cs:25]: Parameters are documented here, but are not "physically" in the constructor.

**Comment by alexatrhea** [comment]
Agreed with @lxatrhea 's proposed changes.

File: CDP4Reporting/Behaviours/ReportDesignerBehaviour.cs (hunk 11 above)
Comment (adrianchivu) [reply CDP4Reporting/Behaviours/ReportDesignerBehaviour.cs]: Done

**Comment by cozminvelciu** [comment]
The commits from the data source branch should not be here as well; those changes should have been merged into the `feat/reporting` branch and then merge that branch here - so that the changes don't appear in the diff for this PR review (and possibly overwrite them!).

After the reporting !492 PR is merged, all files that are extraneous here should be removed; for example, I see some files that do not belong here:
* `CDP4Composition/Reporting/ICDP4ObjectDataSource.cs`
* `CDP4Composition/Reporting/IReportViewModel.cs`
* etc.

**Comment by adrianchivu** [comment]
> The commits from the data source branch should not be here as well; those changes should have been merged into the `feat/reporting` branch and then merge that branch here - so that the changes don't appear in the diff for this PR review (and possibly overwrite them!).
> 
> After the reporting !492 PR is merged, all files that are extraneous here should be removed; for example, I see some files that do not belong here:
> 
> * `CDP4Composition/Reporting/ICDP4ObjectDataSource.cs`
> * `CDP4Composition/Reporting/IReportViewModel.cs`
> * etc.

File: CDP4Composition/Reporting/ICDP4ObjectDataSource.cs [hunk 19]
```diff
@@ -0,0 +1,9 @@
+﻿using CDP4Common.EngineeringModelData;
+
+namespace CDP4Composition.Reporting
+{
+    public interface ICDP4ObjectDataSource
```
Comment (alexatrhea) [inline CDP4Composition/Reporting/ICDP4ObjectDataSource.cs:5]: File missing header and docs

File: CDP4Composition/Reporting/IReportViewModel.cs [hunk 20]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Collections.Generic;
+
+namespace CDP4Composition.Reporting
+{
+    public interface IReportViewModel<T> where T : IReportViewModel<T>
```
Comment (alexatrhea) [inline CDP4Composition/Reporting/IReportViewModel.cs:5]: File missing header and docs

File: CDP4Composition/Reporting/ICDP4ObjectDataSource.cs (hunk 19 above)
Comment (cozminvelciu) [reply CDP4Composition/Reporting/ICDP4ObjectDataSource.cs:5]: These files should not even be here any more, as they are leftovers from the initial demo implementation, which are no longer relevant.

File: CDP4Composition/Reporting/IReportViewModel.cs (hunk 20 above)
Comment (cozminvelciu) [reply CDP4Composition/Reporting/IReportViewModel.cs:5]: These files should not even be here any more, as they are leftovers from the initial demo implementation, which are no longer relevant.

File: CDP4Composition/Adapters/RibbonAdapter.cs [hunk 21]
```diff
@@ -1,6 +1,25 @@
 ﻿// --------------------------------------------------------------------------------------------------------------------
 // <copyright file="RibbonAdapter.cs" company="RHEA System S.A.">
-//   Copyright (c) 2015-2020 RHEA System S.A.
+//    Copyright (c) 2015-2020 RHEA System S.A.
+//
+//    Author: Sam Gerené, Alex Vorobiev, Naron Phou, Patxi Ozkoidi, Alexander van Delft, Mihail Militaru, Nathanael Smiechowski.
```
Comment (cozminvelciu) [inline CDP4Composition/Adapters/RibbonAdapter.cs:5]: Should we add ourselves to the authors?

File: CDP4Composition/Mvvm/Behaviours/RibbonCategoryBehavior.cs [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+//    but WITHOUT ANY WARRANTY; without even the implied warranty of
+//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//    GNU Affero General Public License for more details.
+//
+//    You should have received a copy of the GNU Affero General Public License
+//    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+using System.Diagnostics.CodeAnalysis;
```
Comment (cozminvelciu) [inline CDP4Composition/Mvvm/Behaviours/RibbonCategoryBehavior.cs:26]: Should using be inside namespace?

File: CDP4Reporting.Tests/CDP4Reporting.Tests.csproj [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+    <Reference Include="System.Transactions" />
+    <Reference Include="System.Data.DataSetExtensions" />
+    <Reference Include="Microsoft.CSharp" />
+    <Reference Include="WindowsBase" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\CDP4Composition\CDP4Composition.csproj" />
+    <ProjectReference Include="..\CDP4Reporting\CDP4Reporting.csproj" />
+  </ItemGroup>
+</Project>
```
Comment (cozminvelciu) [inline CDP4Reporting.Tests/CDP4Reporting.Tests.csproj:76]: Some files are missing newlines at end of file.

Comment (adrianchivu) [reply CDP4Reporting.Tests/CDP4Reporting.Tests.csproj:76]: Yes, but I thing this rule is applied only to the source code files.

File: CDP4Composition/Adapters/RibbonAdapter.cs (hunk 21 above)
Comment (adrianchivu) [reply CDP4Composition/Adapters/RibbonAdapter.cs:5]: I think no, at least for this moment.

File: CDP4Reporting.Tests/CDP4Reporting.Tests.csproj (hunk 23 above)
Comment (lxatrhea) [reply CDP4Reporting.Tests/CDP4Reporting.Tests.csproj:76]: @adrianchivu @cozminvelciu That is correct.

File: CDP4Composition/Adapters/RibbonAdapter.cs (hunk 21 above)
Comment (lxatrhea) [reply CDP4Composition/Adapters/RibbonAdapter.cs:5]: You are allowed to do that.

Comment (adrianchivu) [reply CDP4Composition/Adapters/RibbonAdapter.cs:5]: Done
